File: python/search_suggestions_system.py
class Solution:
    def suggestedProducts(
        self, products: list[str], searchWord: str
    ) -> list[list[str]]:
        result: list[int] = list()
        products.sort()
        l: int = 0
        r: int = len(products) - 1
        i: int = 0
        for i in range(len(searchWord)):
            c = searchWord[i]
            while l <= r and (len(products[l]) <= i or products[l][i] != c):
                l += 1
            while l <= r and (len(products[r]) <= i or products[r][i] != c):
                r -= 1

            currLen: int = r - l + 1
            if currLen >= 3:
                result.append(products[l : l + 3])
            else:
                result.append(products[l : r + 1])

        return result


def test_suggestedProducts_case_1():
    # arrange
    products: list[str] = ["mobile", "mouse", "moneypot", "monitor", "mousepad"]
    searchWord: str = "mouse"
    expected: list[str] = [
        ["mobile", "moneypot", "monitor"],
        ["mobile", "moneypot", "monitor"],
        ["mouse", "mousepad"],
        ["mouse", "mousepad"],
        ["mouse", "mousepad"],
    ]

    # act
    solution = Solution()
    actual = solution.suggestedProducts(products, searchWord)

    # assert
    assert expected == actual


def test_suggestedProducts_case_2():
    # arrange
    products: list[str] = ["havana"]
    searchWord: str = "havana"
    expected: list[str] = [
        ["havana"],
        ["havana"],
        ["havana"],
        ["havana"],
        ["havana"],
        ["havana"],
    ]

    # act
    solution = Solution()
    actual = solution.suggestedProducts(products, searchWord)

    # assert
    assert expected == actual


def test_suggestedProducts_case_3():
    # arrange
    products: list[str] = ["bags", "baggage", "banner", "box", "cloths"]
    searchWord: str = "bags"
    expected: list[str] = [
        ["baggage", "bags", "banner"],
        ["baggage", "bags", "banner"],
        ["baggage", "bags"],
        ["bags"],
    ]

    # act
    solution = Solution()
    actual = solution.suggestedProducts(products, searchWord)

    # assert
    assert expected == actual


def test_suggestedProducts_case_4():
    # arrange
    products: list[str] = ["havana"]
    searchWord: str = "tatiana"
    expected: list[str] = [[], [], [], [], [], [], []]

    # act
    solution = Solution()
    actual = solution.suggestedProducts(products, searchWord)

    # assert
    assert expected == actual

File: rust/src/code/search_suggestions_system.rs
struct Solution {}

impl Solution {
    // NOTE: time complexity: O(nlogn + m + n), where n is the length of products and m is the
    // length of search_word.
    pub fn suggested_products(mut products: Vec<String>, search_word: String) -> Vec<Vec<String>> {
        let mut result: Vec<Vec<String>> = Vec::new();
        products.sort();
        let search_word: Vec<char> = search_word.chars().collect();
        let products: Vec<Vec<char>> = products.into_iter().map(|s| s.chars().collect()).collect();
        let mut l: usize = 0;
        let mut r: usize = products.len() - 1;
        for i in 0..search_word.len() {
            let c: char = search_word[i];
            while l <= r && (products[l].len() <= i || products[l][i] != c) {
                l += 1;
            }
            // if r.checked_sub(1).is_none() {
            //     break;
            // }
            while l <= r && (products[r].len() <= i || products[r][i] != c) {
                r -= 1;
            }
            let curr_len: usize = r - l + 1;
            if curr_len >= 3 {
                result.push(
                    products[l..l + 3]
                        .iter()
                        .map(|v| v.iter().collect())
                        .collect(),
                );
            } else {
                result.push(
                    products[l..r + 1]
                        .iter()
                        .map(|v| v.iter().collect())
                        .collect(),
                );
            }
        }

        result
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_suggested_products_case_1() {
        // arrange
        let products: Vec<String> = vec![
            "mobile".to_string(),
            "mouse".to_string(),
            "moneypot".to_string(),
            "monitor".to_string(),
            "mousepad".to_string(),
        ];
        let search_word: String = "mouse".to_string();
        let expected: Vec<Vec<String>> = vec![
            vec![
                "mobile".to_string(),
                "moneypot".to_string(),
                "monitor".to_string(),
            ],
            vec![
                "mobile".to_string(),
                "moneypot".to_string(),
                "monitor".to_string(),
            ],
            vec!["mouse".to_string(), "mousepad".to_string()],
            vec!["mouse".to_string(), "mousepad".to_string()],
            vec!["mouse".to_string(), "mousepad".to_string()],
        ];

        // act
        let actual = Solution::suggested_products(products, search_word);

        // assert
        assert_eq!(expected, actual);
    }

    #[test]
    fn test_suggested_products_case_2() {
        // arrange
        let products: Vec<String> = vec!["havana".to_string()];
        let search_word: String = "havana".to_string();
        let expected: Vec<Vec<String>> = vec![
            vec!["havana".to_string()],
            vec!["havana".to_string()],
            vec!["havana".to_string()],
            vec!["havana".to_string()],
            vec!["havana".to_string()],
            vec!["havana".to_string()],
        ];

        // act
        let actual = Solution::suggested_products(products, search_word);

        // assert
        assert_eq!(expected, actual);
    }
}
